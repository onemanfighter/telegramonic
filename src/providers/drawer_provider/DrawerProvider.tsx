import { DrawerComponent } from '@components';
import { DrawerOpenState, appStore } from '@store';
import { drawerSelector, useShallow } from '@selectors';
import { useDisclosure } from '@chakra-ui/react';
import { useEffect } from 'react';
import { DrawerProviderProps } from './types';

const DrawerProvider = ({ children }: DrawerProviderProps) => {
  const { isOpen, onOpen, onClose } = useDisclosure();
  const { openState, closeDrawer, drawerData } = appStore(
    useShallow(drawerSelector),
  );
  const onModalClose = () => {
    onClose();
    closeDrawer();
  };
  useEffect(() => {
    if (openState === DrawerOpenState.OPEN) {
      onOpen();
    } else {
      onModalClose();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [openState, onOpen]);

  return (
    <>
      <DrawerComponent
        {...drawerData}
        onModalClose={onModalClose}
        isOpen={isOpen}
      />
      {children}
    </>
  );
};

export default DrawerProvider;
