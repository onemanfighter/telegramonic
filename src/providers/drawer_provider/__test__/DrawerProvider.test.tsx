import {
  act,
  fireEvent,
  render,
  renderHook,
  screen,
} from '@testing-library/react';
import { AlertSelector, drawerSelector, useShallow } from '@selectors';
import { AlertComponentState, DrawerState, appStore } from '@store';
import DrawerProvider from '../DrawerProvider';

jest.useFakeTimers();

describe('DrawerProvider', () => {
  const onModalCloseMock = jest.fn();

  const drawerState: DrawerState['drawerData'] = {
    isOpen: false,
    title: 'Drawer',
    children: <div>Footer</div>,
    footer: <div>Footer</div>,
    onModalClose: onModalCloseMock,
  };
  beforeEach(() => {
    onModalCloseMock.mockClear();
  });

  it('should render children', () => {
    const { container } = render(
      <DrawerProvider>
        <div>child</div>
      </DrawerProvider>,
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByText('child')).toBeDefined();
  });

  it('should open drawer component on calling open drawer with drawer state', () => {
    const { result } = renderHook(() => appStore(useShallow(drawerSelector)));
    const { container } = render(
      <DrawerProvider>
        <div>child</div>
      </DrawerProvider>,
    );

    act(() => {
      result.current.openDrawer(drawerState);
    });

    expect(container).toMatchSnapshot();
    expect(screen.getByText('Drawer')).toBeDefined();
  });

  it.skip('should hide the drawer on clicking close button', async () => {
    const { result } = renderHook(() => appStore(useShallow(AlertSelector)));
    const { container } = render(
      <DrawerProvider>
        <div>child</div>
      </DrawerProvider>,
    );

    fireEvent.click(screen.getByLabelText('close-button'));

    act(async () => {
      jest.runAllTimers();
    });

    expect(onModalCloseMock).toHaveBeenCalledTimes(1);

    expect(container).toMatchSnapshot();
    expect(screen.queryByText('Drawer')).toBeNull();
  });
});
