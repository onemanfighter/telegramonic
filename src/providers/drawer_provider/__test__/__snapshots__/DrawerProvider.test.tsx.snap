// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerProvider should hide the drawer on clicking close button 1`] = `
<div>
  <div>
    child
  </div>
</div>
`;

exports[`DrawerProvider should open drawer component on calling open drawer with drawer state 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
>
  <div>
    child
  </div>
</div>
`;

exports[`DrawerProvider should render children 1`] = `
<div>
  <div>
    child
  </div>
</div>
`;
